gr-digital/lib/mimo_channel_estimator_cc_impl.cc

```diff
@@ -0,0 +1,246 @@
+          break;
+        }
+        default: {
+#if (EIGEN3_ENABLED)
+          // TODO calculate pseudo-inverse in constructor
+          // Map training sequence and received sequence 2-dimensional std::vector to Eigen MatrixXcf.
+          Eigen::MatrixXcf training_matrix(d_M, d_training_length);
+          Eigen::MatrixXcf received_training_matrix(d_N, d_training_length);
+          for (int i = 0; i < d_M; ++i) {
+              training_matrix.row(i) = Eigen::VectorXcf::Map(&d_training_sequence[i][0], d_training_length);
+          }
+          for (int i = 0; i < d_N; ++i) {
+              received_training_matrix.row(i) = Eigen::VectorXcf::Map((const gr_complex *) input_items[i], d_training_length);
+          }
+          // Calculate the Maximum Likelihood estimation for the channel matrix.
+          Eigen::MatrixXcf csi = received_training_matrix * training_matrix.adjoint() * (training_matrix*training_matrix.adjoint()).inverse();
+
+           // Map the CSI Eigen MatrixXcf to a 2-dim std::vector.
+          for (int i = 0; i < d_N; ++i) {
+            Eigen::VectorXcf::Map(&d_csi[i][0], d_M) = csi.row(i);
+          }
+#else
+          throw std::runtime_error("Required library Eigen3 for MxN MIMO schemes with M,N>2 not installed.");
+#endif
+        }
+      }
+    }
+
+    int
+    mimo_channel_estimator_cc_impl::general_work (int noutput_items,
+                       gr_vector_int &ninput_items,
+                       gr_vector_const_void_star &input_items,
+                       gr_vector_void_star &output_items)
+    {
+      // Collect all tags of the input buffer with key "csi" in the vector 'tags'.
+      get_tags_in_window(tags, 0, 0, noutput_items, d_key);
+
+      uint32_t nconsumed = 0; // Number of read items.
+      uint32_t nwritten = 0; // Number of written items.
+      uint32_t symbol_length; // Number of items in the current symbol.
+
+      if(tags.size() == 0){ // Input buffer includes no tags at all.
+        // Handle all samples in buffer as they belong to the current symbol.
+        symbol_length = noutput_items;
+        // Copy the data to the output buffer.
+        copy_symbols(input_items, output_items, symbol_length, 0, 0);
+        nconsumed = noutput_items;
+        nwritten = noutput_items;
+      } else { // Input buffer includes tags.
+        if (tags[0].offset - nitems_read(0) > 0){
+          /* There are items in the input buffer, before the first tag arrives,
++          * which belong to the previous symbol. */
+          symbol_length = tags[0].offset - nitems_read(0);
+          // Copy the data to the output buffer.
+          copy_symbols(input_items, output_items, symbol_length, 0, 0);
+          nconsumed += symbol_length;
+          nwritten += symbol_length;
+        }
+        // Iterate over tags in buffer.
+        for (unsigned int i = 0; i < tags.size(); ++i) {
+          // Calculate the number of items before the next tag.
+          if (i < tags.size() - 1) {
```

Comment:
Why not simply split into a `for` loop with condition `i < tags.size() - 1` and a separate `i = tags.size()-1`?

That'd enhance readability.